src/data_race.rs

```diff
@@ -0,0 +1,1189 @@
+            let data_race = &alloc_meta.global;
+            data_race.maybe_perform_sync_operation(|index, mut clocks| {
+                for (_,range) in alloc_meta.alloc_ranges.borrow_mut().iter_mut(place_ptr.offset, size) {
+                    if let Err(DataRace) = op(range, &mut *clocks, index, atomic) {
+                        mem::drop(clocks);
+                        return VClockAlloc::report_data_race(
+                            &alloc_meta.global, range, description, true,
+                            place_ptr, size
+                        );
+                    }
+                }
+                Ok(())
+            })?;
+
+            // Log changes to atomic memory
+            if log::log_enabled!(log::Level::Trace) {
+                for (_,range) in alloc_meta.alloc_ranges.borrow().iter(place_ptr.offset, size) {
+                    log::trace!(
+                        "Updated atomic memory({:?}, offset={}, size={}) to {:#?}",
+                        place.ptr.assert_ptr().alloc_id, place_ptr.offset.bytes(), size.bytes(),
+                        range.atomic_ops
+                    );
+                }
+            }
+        }
+        Ok(())
+    }
+
+}
+
+/// Handle for locks to express their
+///  acquire-release semantics
+#[derive(Clone, Debug, Default)]
+pub struct DataRaceLockHandle {
+
+    /// Internal acquire-release clock
+    ///  to express the acquire release sync
+    ///  found in concurrency primitives
+    clock: VClock,
+}
+impl DataRaceLockHandle {
+    pub fn set_values(&mut self, other: &Self) {
+        self.clock.clone_from(&other.clock)
+    }
+    pub fn reset(&mut self) {
+        self.clock.set_zero_vector();
+    }
+}
+
+
+/// Error returned by finding a data race
+///  should be elaborated upon
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub struct DataRace;
+
+/// Externally stored memory cell clocks
+///  explicitly to reduce memory usage for the
+///  common case where no atomic operations
+///  exists on the memory cell
+#[derive(Clone, PartialEq, Eq, Default, Debug)]
+struct AtomicMemoryCellClocks {
+
+    /// The clock-vector for the set of atomic read operations
+    ///  used for detecting data-races with non-atomic write
+    ///  operations
+    read_vector: VClock,
+
+    /// The clock-vector for the set of atomic write operations
+    ///  used for detecting data-races with non-atomic read or
+    ///  write operations
+    write_vector: VClock,
+
+    /// Synchronization vector for acquire-release semantics
+    ///   contains the vector of timestamps that will
+    ///   happen-before a thread if an acquire-load is 
+    ///   performed on the data
+    sync_vector: VClock,
+
+    /// The Hash-Map of all threads for which a release
+    ///  sequence exists in the memory cell, required
+    ///  since read-modify-write operations do not
+    ///  invalidate existing release sequences 
+    release_sequences: VSmallClockSet,
```

Comment:
I am somewhat surprised that release sequences need special treatment here... I seem to recall having seen a "clock-based model" where they just fall out without special treatment. But I might misremember. I'll have to dig deeper into that paper you are implementing it seems.